// humane_list/lib/index.js
// Generated by CoffeeScript 1.3.3
(function() {
  var Humane_List,
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  Humane_List = (function() {
    var Element;

    Humane_List.prototype.Element = Element = (function() {

      function Element() {
        this.d = {};
        switch (arguments.length) {
          case 1:
            this.d.keys = [];
            this.d.val = arguments[0];
            break;
          case 2:
            this.d.keys = arguments[0];
            this.d.val = arguments[1];
            break;
          case 3:
            this.d.pos = arguments[0];
            this.d.keys = arguments[1];
            this.d.val = arguments[2];
            break;
          default:
            throw new Error("Unknown quantity of arguments: " + arguments);
        }
        this.d.keys = this.d.keys.shift ? this.d.keys : [this.d.keys];
      }

      Element.prototype.keys = function() {
        return this.d.keys;
      };

      Element.prototype.value = function() {
        return this.d.val;
      };

      Element.prototype.position = function() {
        return this.d.pos;
      };

      Element.prototype.update_position = function(n) {
        return this.d.pos = n;
      };

      Element.prototype.to_array = function() {
        return [this.d.keys, this.d.val];
      };

      Element.prototype.remove_alias = function(alias) {
        var k;
        return this.d.keys = (function() {
          var _i, _len, _ref, _results;
          _ref = this.d.keys;
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            k = _ref[_i];
            if (k !== alias) {
              _results.push(k);
            }
          }
          return _results;
        }).call(this);
      };

      return Element;

    })();

    function Humane_List(vals) {
      var k, v, _i, _len;
      this.d = {};
      this.d.core = [];
      if (vals && vals.shift) {
        for (_i = 0, _len = vals.length; _i < _len; _i++) {
          v = vals[_i];
          this.push('end', v);
        }
      } else if (vals) {
        for (k in vals) {
          v = vals[k];
          this.push('end', k, v);
        }
      }
    }

    Humane_List.prototype.keys = function() {
      return this.d.keys;
    };

    Humane_List.prototype.pop = function(pos) {
      var ele;
      ele = pos.toString() === 'front' ? this.d.core.shift() : this.d.core.pop();
      if (!ele) {
        return ele;
      }
      return ele.value();
    };

    Humane_List.prototype.push = function() {
      var args, e, i, last, last_v, num_pos, pos, v, _i, _len, _ref;
      args = arguments;
      pos = args[0];
      num_pos = typeof pos === "number" ? pos : pos === 'front' ? 1 : (last = this.positions().pop(), (last && (last + 1)) || this.d.core.length + 1);
      switch (args.length) {
        case 2:
          e = num_pos ? new this.Element(num_pos, [], args[1]) : new this.Element(args[1]);
          break;
        case 3:
          e = num_pos ? new this.Element(num_pos, args[1], args[2]) : new this.Element(args[1], args[2]);
          break;
        default:
          throw new Error("Invalid arguments: " + arguments);
      }
      this.d.core.unshift(e);
      this.d.core = this.d.core.sort(function(a, b) {
        return a.position() > b.position();
      });
      if (this.d.core.length > 1) {
        _ref = this.d.core;
        for (i = _i = 0, _len = _ref.length; _i < _len; i = ++_i) {
          last_v = _ref[i];
          v = this.d.core[i + 1];
          if (v) {
            if (last_v.position() >= v.position()) {
              v.update_position(v.position() + 1);
            }
          }
        }
      }
      return e.position();
    };

    Humane_List.prototype.front = function() {
      return this.d.core[0] && this.d.core[0].value();
    };

    Humane_List.prototype.end = function() {
      var row;
      row = this.d.core[this.d.core.length - 1];
      return row && row.value();
    };

    Humane_List.prototype.at_key = function(k) {
      var ele, v, _i, _len, _ref;
      _ref = this.d.core;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        ele = _ref[_i];
        if (__indexOf.call(ele.keys(), k) >= 0) {
          v = ele.value();
        }
      }
      return v;
    };

    Humane_List.prototype.at_position = function(n) {
      var comp_pos;
      comp_pos = this.to_key_or_computer_position(n);
      return this.d.core[comp_pos] && this.d.core[comp_pos].value();
    };

    Humane_List.prototype.has_key = function(k) {
      var found, v;
      found = (function() {
        var _i, _len, _ref, _results;
        _ref = this.d.core;
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          v = _ref[_i];
          if (__indexOf.call(v.keys(), k) >= 0) {
            _results.push(v.keys());
          }
        }
        return _results;
      }).call(this);
      return found.length > 0;
    };

    Humane_List.prototype.positions = function() {
      var v, _i, _len, _ref, _results;
      _ref = this.d.core;
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        v = _ref[_i];
        _results.push(v.position());
      }
      return _results;
    };

    Humane_List.prototype.keys = function() {
      var ele, _i, _len, _ref, _results;
      _ref = this.d.core;
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        ele = _ref[_i];
        _results.push(ele.keys());
      }
      return _results;
    };

    Humane_List.prototype.values = function() {
      var v, _i, _len, _ref, _results;
      _ref = this.d.core;
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        v = _ref[_i];
        _results.push(v.value());
      }
      return _results;
    };

    Humane_List.prototype.concat = function(pos, o) {
      var i, key, keys, positions, v, val, _i, _j, _len, _len1, _results, _results1;
      if (o.shift) {
        positions = (function() {
          var _i, _len, _results;
          _results = [];
          for (i = _i = 0, _len = o.length; _i < _len; i = ++_i) {
            v = o[i];
            _results.push(i);
          }
          return _results;
        })();
        if (pos === 'front') {
          positions = positions.reverse();
        }
        _results = [];
        for (_i = 0, _len = positions.length; _i < _len; _i++) {
          i = positions[_i];
          _results.push(this.push(pos, o[i]));
        }
        return _results;
      } else {
        keys = (function() {
          var _results1;
          _results1 = [];
          for (key in o) {
            val = o[key];
            _results1.push(key);
          }
          return _results1;
        })();
        if (pos === 'front') {
          keys = keys.reverse();
        }
        _results1 = [];
        for (_j = 0, _len1 = keys.length; _j < _len1; _j++) {
          key = keys[_j];
          _results1.push(this.push(pos, key, o[key]));
        }
        return _results1;
      }
    };

    Humane_List.prototype.to_key_or_computer_position = function(key_or_human_pos) {
      if (typeof key_or_human_pos === 'number') {
        return key_or_human_pos - 1;
      }
      return key_or_human_pos;
    };

    Humane_List.prototype.get_computer_position = function(key_or_human_pos) {
      var k, key_or_pos, pos, v, _i, _len, _ref;
      key_or_pos = this.to_key_or_computer_position(key_or_human_pos);
      if (!key_or_pos) {
        return key_or_pos;
      }
      _ref = this.d.core;
      for (k = _i = 0, _len = _ref.length; _i < _len; k = ++_i) {
        v = _ref[k];
        if ((k === key_or_pos) || (__indexOf.call(v.keys(), key_or_pos) >= 0)) {
          pos = k;
        }
      }
      return pos;
    };

    Humane_List.prototype.get_computer_position_or_throw = function(key_or_human_pos) {
      var pos;
      pos = this.get_computer_position(key_or_human_pos);
      if (!pos) {
        throw new Error("Key/pos is not defined: " + key_or_human_pos);
      }
      return pos;
    };

    Humane_List.prototype.alias = function(key_or_pos, nickname) {
      var keys, pos;
      pos = this.get_computer_position_or_throw(key_or_pos);
      keys = this.d.core[pos].keys();
      if (__indexOf.call(keys, nickname) >= 0) {
        return nickname;
      }
      return keys.push(nickname);
    };

    Humane_List.prototype.remove_alias = function(nickname) {
      var ele, _i, _len, _ref;
      _ref = this.d.core;
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        ele = _ref[_i];
        ele.remove_alias(nickname);
      }
      return nickname;
    };

    Humane_List.prototype.delete_at = function(target_k) {
      var k, old, pos, row, v;
      pos = this.get_computer_position(target_k);
      row = this.d.core[pos];
      if (!row) {
        return row;
      }
      old = this.at_position(pos);
      this.d.core = (function() {
        var _i, _len, _ref, _results;
        _ref = this.d.core;
        _results = [];
        for (k = _i = 0, _len = _ref.length; _i < _len; k = ++_i) {
          v = _ref[k];
          if (k !== pos) {
            _results.push(v);
          }
        }
        return _results;
      }).call(this);
      return old;
    };

    return Humane_List;

  })();

  exports.Humane_List = Humane_List;

}).call(this);
